### ISSUE TITLE
v3.3.1 Changes

### ISSUE THREAD

**Issue by majora2007** [vers-one/EpubReader#93]
Opening this thread so I can ask some questions as I migrate the breaking changes.

First: I understand we no longer need to iterate through keys to validate, however, how can I check if a key is valid or not? 

Previously:
`if(!book.Content.Images.ContainsKey(key)) continue`

Now, we only have GetLocalFileRefByKey which doesn't return null per annotations, so not sure how to proceed.

Second: I'm trying to figure out how to rewrite the following code now that both filepath and key are available. For example, this code here gets keys from images in the document html and tries to map it back to the image file. It of course has hacks for the common ../path issue.
```
private static string GetKeyForImage(EpubBookRef book, string imageFile)
    {
        if (book.Content.Images.Local.ContainsKey(imageFile)) return imageFile;

        var correctedKey = book.Content.Images.Local.Keys.SingleOrDefault(s => s.EndsWith(imageFile));
        if (correctedKey != null)
        {
            imageFile = correctedKey;
        }
        else if (imageFile.StartsWith(".."))
        {
            // There are cases where the key is defined static like OEBPS/Images/1-4.jpg but reference is ../Images/1-4.jpg
            correctedKey =
                book.Content.Images.Local.Keys.SingleOrDefault(s => s.EndsWith(imageFile.Replace("..", string.Empty)));
            if (correctedKey != null)
            {
                imageFile = correctedKey;
            }
        }


        return imageFile;
    }
```

Third, how can I handle this situation:

```
// Try to get the cover image from OPF file, if not set, try to parse it from all the files, then result to the first one.
            var coverImageContent = epubBook.Content.Cover
                                    ?? epubBook.Content.Images.Local.Values.FirstOrDefault(file => Parser.IsCoverImage(file.FilePath)) // FileName -> FilePath
                                    ?? epubBook.Content.Images.Local.Values.FirstOrDefault();
```

In this situation, I need to try and grab Cover and if not defined, fallback to the first Image in the epub. 

Would this suffice?
```
var coverImageContent = epubBook.Content.Cover
                                    ?? epubBook.Content.Images.Local.FirstOrDefault(file => Parser.IsCoverImage(file.FilePath)) // FileName -> FilePath
                                    ?? epubBook.Content.Images.Local.FirstOrDefault();
```

**Comment by vers-one** [comment]
1. `if(!book.Content.Images.ContainsKey(key)) continue` &rarr; `if(!book.Content.Images.ContainsLocalFileRefWithKey(key)) continue` (assuming that `book` is `EpubBookRef`).
2. It depends on what value is passed as `imageFile` and how the return value of this method is used. Typically, applications don't need to map absolute file paths to file keys, unless they access manifest items in the schema. Could you post the link to the code where this method is used?
3. Yes, this replacement is correct. Although, there could be a way to do this without iteration. Could you post the link to the `Parser.IsCoverImage` method?

**Comment by majora2007** [comment]
Thanks for the quick reply, let me dive in and see what I can do. 2 big breaking changes in a row, you're killing me ;)

**Comment by vers-one** [comment]
Yeah, in hindsight, I should have made all these design changes when I added EPUB 3 support in 2019. At that time I hoped that I will be able to map EPUB 3 data structures to the EpubReader types designed for EPUB 2 but unfortunately, it didn't work out.
